source/common/router/router.cc

```diff
@@ -502,7 +501,16 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
   route_entry_->finalizeRequestHeaders(headers, callbacks_->streamInfo(),
                                        !config_.suppress_envoy_headers_);
-  FilterUtility::setUpstreamScheme(headers, *cluster_);
+  envoy::api::v2::core::Metadata host_metadata;
```

Comment:
rather than resolving transportSocketFactory again, can this get from the client connection from ConnPool? (having a flag in conn pool seems fine)

Reply:
The transport socket factory is actually not resolved during the pool selection call chain. Right now it only occurs as part of the connection event, when upstream is about to make connection. At this point, no client connection is being created.

1 pool is corresponding to 1 host (selected), I could add a flag to conn pool. This would require to plumb/add `ClusterInfo` reference in [ClusterManagerImpl](https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/source/common/upstream/cluster_manager_impl.cc#L1269)  to the conn pool constructor though, since resolving socket factory requires

- `ClusterInfo::transport_socket_match_`
- host::metadata (which the pool has as part of the constructor already.
- And then we set a constant member field into `HttpConnectionPool`, and able to return with a accessor method.

Does above sound good to you?

Reply:
Hmm, I thought the host selection resolves the transport socket factory, no?

Reply:
Actually, I just realized the Host itself already has access to ClusterInfo, plus it owns metadata and address as well, so `HostDescription` itself can just provide this information.

But extra invocation of the `TransportSocketMatcher::resolve` could mess up with the stats though. maybe add extra param like `bool create_connection`, and only increment stats when `create_connection` is true?

Reply:
Yeah HostDescription seems a perfect place, as well as to do `TransportSocketMatcher::resolve`. I don't think you should mess stats and the `bool create_connection` flag, currently the transport socket resolving is pretty static so don't even have to be per-connection basis, it can be only done once per `Host/HostDescription`, right?

Reply:
that's right.

> transport socket resolving is pretty static 

sounds good. I can add `TransporSocketMatcher::statsByHost()` and then HostDescription constructor will be responsible for incrementing that stats entry.

The counter will be per host basis.

Reply:
sgtm

Reply:
I end up adding `transportSocketFactory()` for the `HostDesription`. I just realize we shouldn't go through match loop iteration every time we make a connection. Instead we can just do it once, avoiding the computation.

I think this is okay, even the CDS push updates occur, Envoy should recycle those old Host object, right?

Reply:
> I think this is okay, even the CDS push updates occur, Envoy should recycle those old Host object, right?

If cluster didn't change yes for sure, not sure in the case if cluster changed, though you'll need to recalculate the match so it doesn't matter either.